### PR TITLE
Synthesizing tweets LSTM model and relevant data file

### PR DESCRIPTION
**Pull request by rhiag**
Naming: feature/descriptive-name

Reviewers: minimum 2 (Chris and/or 2 from Web or DS respectively and not person who submitted PR)

Description: 
  What is the motivation for changes?
  What specific changes were made?

Submitter Checklist: 
  - [x] Small scope (1-2 components)
  - [x] Not too many concerns in a single commit
  - [ ] Remove extraneous comments (code in comments), print statements, to-dos, console.logs, extra fluff
  - [ ] Atomic, descriptive commit messages
  - [ ] Consistent formatting

Reviewer Checklist:
  - [ ] Small scope (1-2 components)
  - [ ] Comments, print statements, to-dos, console.logs, extra fluff were removed
  - [ ] Give feedback no matter what conversations happened elsewhere, document here
  - [ ] Comment on the line of code itself (blue plus box) in files changed tab

### CODE REVIEW

**Comment by rhiag** [comment]
This pull request is still a work in progress. 
I have attempted to train a model on data for labels 2 and 3 incident reports. 
I still need to check if the tweets being generated are relevant.

**Review by JAaron93 — COMMENTED** [review]
Looks good for a draft pull request.

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb [hunk 1]
```diff
@@ -0,0 +1,1225 @@
...
+            "Epoch 123/340\n",
+            "146/146 [==============================] - 32s 216ms/step - loss: 0.1104 - accuracy: 0.9859\n",
+            "Epoch 124/340\n",
+            "146/146 [==============================] - 31s 216ms/step - loss: 0.0914 - accuracy: 0.9892\n",
+            "Epoch 125/340\n",
+            "146/146 [==============================] - 32s 216ms/step - loss: 0.0790 - accuracy: 0.9911\n",
+            "Epoch 126/340\n",
+            "146/146 [==============================] - 32s 218ms/step - loss: 0.0712 - accuracy: 0.9923\n",
+            "Epoch 127/340\n",
+            "146/146 [==============================] - 32s 218ms/step - loss: 0.0636 - accuracy: 0.9932\n",
```
Comment (BOsterbuhr) [inline notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:980]: Do you think this would be a good cutoff point? This training looks awesome, excuse my ignorance but what exactly is this training?

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb [hunk 2]
```diff
@@ -0,0 +1,1225 @@
...
+          "name": "stderr"
+        },
+        {
+          "output_type": "execute_result",
+          "data": {
+            "application/vnd.google.colaboratory.intrinsic+json": {
+              "type": "string"
+            },
+            "text/plain": [
+              "'block of sw alder st one officer strikes a protester repeatedly with their baton that protester was later arrested at an antiice protest in the portland pearl district several police'"
```
Comment (BOsterbuhr) [inline notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:1203]: Did the model write this line?

**Review by BOsterbuhr — COMMENTED** [review]
This looks promising, is the purpose of this to generate rank 2 and 3 incidents?

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb (hunk 1 above)
Comment (rhiag) [reply notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:980]: The purpose of the model is to generate rank 2 and 3 tweets. At this point I have the model running for 2 different scenarios:
1. on training data which has both 2 and 3 ranked incidents 
2. a model running on label 2 and label 3 separately.
The thought process behind it is to see which would give us better results. However, I suspect that since the data for the two labels is relatively small we may do better on the combined data. 
Do any of you have thoughts on that?

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb (hunk 2 above)
Comment (rhiag) [reply notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:1203]: It did generate that line. Going by what I'm seeing as results, I wonder if we need perfect sentences to serve as input to the BERT model. If so then there is either more tweaking on this LSTM to be done or manually just grabbing snippets of the generated tweets.

Comment (mbenbanaste) [reply notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:1203]: This looks better than I expected it would. I don't think we need "perfect sentences" for BERT; we need "admissible sentences," which is harder to figure out. This generated line is not perfect but one could interpret it as the user hastily first writing the location and moving on to write "one officer strikes a protester..." Tweets (and sentences contained) aren't always composed of perfect sentences, but they obviously need to make sense. What I wonder is whether we could generate hundreds of admissible sentences through this method, so having mass production - and whether (and if so at what point) it would cease to generate admissible sentences

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb (hunk 1 above)
Comment (JAaron93) [reply notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:980]: Are you going to implement early stopping as a callback so it doesn't run for hundreds of epochs?

Which is more accurate? Training on both Rank 2 & 3 simultaneously or training on each rank separately? I would assume a priori that training on Rank 2 & 3 separately would give us better results for each.

**Comment by rhiag** [comment]
An update on this exercise:
After a closer inspection of the tweets that were generated using this model and having a conversation with Joshua about it, I realized that these tweets did not really cut it and there would be a lot of manual work still involved. I was ready to begin making these manual changes or look at an alternative solution. However, Joshua has used the GPT2 model and generated some fantastic tweets. 
Should anyone be curious to see take a look at the tweets generated by the LSTM model, I'm happy to send them to you.

**Comment by JAaron93** [comment]
> An update on this exercise:
> After a closer inspection of the tweets that were generated using this model and having a conversation with Joshua about it, I realized that these tweets did not really cut it and there would be a lot of manual work still involved. I was ready to begin making these manual changes or look at an alternative solution. However, Joshua has used the GPT2 model and generated some fantastic tweets.
> Should anyone be curious to see take a look at the tweets generated by the LSTM model, I'm happy to send them to you.

I'm actually still curious about the .csv of LSTM generated tweets. Would be nice to serve as a comparison to GPT-2 generated tweets and why it was needed to use such a generative text model in the first place. Could you upload it for us?

File: notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb (hunk 1 above)
Comment (Andrew-Haney) [reply notebooks/labs37_notebooks/GenerateTweet_LSTM.ipynb:980]: As Joshua mentioned, I think implementing an early stop as a callback to keep it from continuing to run a couple hundred epochs is a fantastic idea!  Other than that great accuracy and very clean, easy to read code.  Good job!

**Review by Andrew-Haney — CHANGES_REQUESTED** [review]
Overall awesome job! I would just implement the early stopping to prevent the model from running any unnecessary epochs.  Great job!

**Comment by rhiag** [comment]
Update: 
We have gone ahead and decided to use GPT-2 pretrained model in order to synthesize tweets as that required much less manual work and was given amazing  results. 
Im unsure if we want to have this notebook on the repo. If so, I think it should live in the archived folder.
